File: compiler/bench/bucketsort.fs
\ expected: 49995000
\ Bucket sort benchmark - sort 10000 elements

10000 constant SIZE
100 constant NBUCKETS
create arr SIZE cells allot
create buckets NBUCKETS SIZE * cells allot
create bcount NBUCKETS cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;
: bcount@ ( i -- n ) cells bcount + @ ;
: bcount! ( n i -- ) cells bcount + ! ;
: bucket@ ( b i -- n ) swap SIZE * + cells buckets + @ ;
: bucket! ( n b i -- ) swap SIZE * + cells buckets + ! ;

: bucket-idx ( val -- b )
  SIZE / ;

: insert-bucket ( val -- )
  dup bucket-idx         \ val b
  dup bcount@            \ val b cnt
  2dup bucket!           \ store val at bucket[b][cnt]
  1+ swap bcount! ;      \ increment count

: sort-bucket ( b -- )
  \ Insertion sort on bucket b
  dup bcount@ 1 do       \ b
    dup i bucket@        \ b key
    i                    \ b key j
    begin
      dup 0> 2 pick 3 pick 2 pick 1- bucket@ < and
    while
      2 pick over 1- bucket@ 2 pick swap bucket!
      1-
    repeat
    2 pick swap bucket!  \ b
    drop
  loop drop ;

: collect-buckets ( -- )
  0                       \ idx
  NBUCKETS 1- 0 swap do   \ descending for descending order
    i bcount@ 0 do
      i j bucket@        \ idx val
      over arr!
      1+
    loop
  -1 +loop drop ;

: bucket-sort ( -- )
  NBUCKETS 0 do 0 i bcount! loop
  SIZE 0 do i arr@ insert-bucket loop
  NBUCKETS 0 do i sort-bucket loop
  collect-buckets ;

: init-arr ( -- )
  SIZE 0 do
    SIZE i - 1- i arr!
  loop ;

: sum-arr ( -- n )
  0 SIZE 0 do i arr@ + loop ;

: main
  init-arr
  bucket-sort
  sum-arr . cr ;

File: compiler/bench/bucketsort.c
// expected: 49995000
// Bucket sort benchmark - sort 10000 elements

#include <stdio.h>

#define SIZE 10000
#define NBUCKETS 100

int arr[SIZE];
int buckets[NBUCKETS][SIZE];
int bcount[NBUCKETS];

int bucket_idx(int val) {
    return val / SIZE;
}

void insert_bucket(int val) {
    int b = bucket_idx(val);
    buckets[b][bcount[b]++] = val;
}

void sort_bucket(int b) {
    // Insertion sort on bucket b
    for (int i = 1; i < bcount[b]; i++) {
        int key = buckets[b][i];
        int j = i;
        while (j > 0 && buckets[b][j - 1] < key) {
            buckets[b][j] = buckets[b][j - 1];
            j--;
        }
        buckets[b][j] = key;
    }
}

void collect_buckets(void) {
    int idx = 0;
    // Descending order
    for (int b = NBUCKETS - 1; b >= 0; b--) {
        for (int i = 0; i < bcount[b]; i++) {
            arr[idx++] = buckets[b][i];
        }
    }
}

void bucket_sort(void) {
    for (int i = 0; i < NBUCKETS; i++) bcount[i] = 0;
    for (int i = 0; i < SIZE; i++) insert_bucket(arr[i]);
    for (int i = 0; i < NBUCKETS; i++) sort_bucket(i);
    collect_buckets();
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = SIZE - i - 1;
    }
}

long sum_arr(void) {
    long sum = 0;
    for (int i = 0; i < SIZE; i++) {
        sum += arr[i];
    }
    return sum;
}

int main(void) {
    init_arr();
    bucket_sort();
    printf("%ld\n", sum_arr());
    return 0;
}
